pkg/pillar/cmd/zedagent/handlentp.go

```diff
@@ -0,0 +1,336 @@
+
+	wdName := agentName + "-ntp"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	ts := time.Now()
+	for {
+		select {
+		case <-ticker.C:
+			dest := ControllerDest
+			if time.Now().Sub(ts) > getForceSendInterval(ctx) {
+				// Periodically do a force send to be sure
+				// controller gets an NTP update in any case
+				dest |= ForceSend
+				ts = time.Now()
+			}
+			publishNTPSources(ctx, wdName, dest)
+		case dest := <-triggerNTPSourcesInfo:
+			publishNTPSources(ctx, wdName, dest)
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func publishNTPSources(ctx *zedagentContext, wdName string,
+	dest destinationBitset) {
+	info := getNTPSourcesInfo(ctx)
+	if info == nil {
+		// Not available.
+		return
+	}
+	start := time.Now()
+	publishNTPSourcesToDest(ctx, info, dest)
+	ctx.ps.CheckMaxTimeTopic(wdName, "publishNTPSources", start,
+		warningTime, errorTime)
+}
+
+func ntpSourcesChanged(info *info.ZInfoNTPSources,
+	dest destinationBitset) bool {
+
+	prev, cached := prevInfoNTPSourcesMap[dest]
+	prevInfoNTPSourcesMap[dest] = info
+	if !cached {
+		// First time
+		return true
+	}
+	if len(prev.Sources) != len(info.Sources) {
+		// Set of sources has been changed
+		return true
+	}
+	for i, source := range info.Sources {
+		prevSource := prev.Sources[i]
+		if prevSource.State != source.State ||
+			prevSource.Mode != source.Mode ||
+			prevSource.Reachable != source.Reachable {
+			// Some fields have been changed
+			return true
+		}
+	}
+	// Same
+	return false
+}
+
+func publishNTPSourcesToDest(ctx *zedagentContext,
+	infoNTPSources *info.ZInfoNTPSources, dest destinationBitset) {
+
+	if dest == 0 {
+		log.Errorf("publishNTPSourcesToDest: incorrect destination")
+		return
+	}
+	if (dest &^ LPSDest) == 0 {
+		// TODO: we don't support LPS
+		return
+	}
+	changed := ntpSourcesChanged(infoNTPSources, dest)
+	if !changed && (dest&ForceSend) == 0 {
+		// Nothing was changed and force send is not requested,
+		// so just return
+		return
+	}
+	infoMsg := &info.ZInfoMsg{
+		Ztype: info.ZInfoTypes_ZiNTPSources,
+		DevId: devUUID.String(),
+		InfoContent: &info.ZInfoMsg_NtpSources{
+			NtpSources: infoNTPSources,
+		},
+		AtTimeStamp: ptypes.TimestampNow(),
+	}
+
+	log.Functionf("publishNTPSourcesToDest: sending %v", infoMsg)
+	data, err := proto.Marshal(infoMsg)
+	if err != nil {
+		log.Fatal("publishNTPSourcesToController: proto marshaling error: ", err)
+	}
+	buf := bytes.NewBuffer(data)
+	if buf == nil {
+		log.Fatal("malloc error")
+	}
+	size := int64(proto.Size(infoMsg))
+
+	const bailOnHTTPErr = false
+	const withNetTrace = false
+	key := "ntpsources:" + devUUID.String()
+
+	// Even for the controller destination we can't stall the queue on error,
+	// set @forcePeriodic to true and let the timer repeat the send in a while
```

Comment:
Doesn't this mean that if the network is not up (or that there is a timeout), you will wait for the full 10 minutes until you try again?

Why can't you have the queue retry on error for the triggered updates (and have the periodic updates not retry)?
How will the queue "stall" in a way which will be more upsetting than e.g., a device INFO message being queued and retried?
(You might want to think about the relative priority of the new NTP INFO message if you didn't already.)

Reply:
> you will wait for the full 10 minutes until you try again?

Yes, if network is not up then we have to wait another 10 minutes. Exactly as all other info messages function (with less interval though). So do not quite understand why NTP status (which has pure informative role on the controller) should be exceptional here.

> Why can't you have the queue retry on error for the triggered updates (and have the periodic updates not retry)?

We don't have any callback mechanism and any error checks and retries. My impression is that this particular problem does not worth any over-complication, at least that what I understood discussing this with Daniel.

> How will the queue "stall" in a way which will be more upsetting than e.g., a device INFO message being queued and retried?

Not upsetting at all, but abuses the semantics chosen long time ago to have two queues: one for events and another one for recurring calls. If you are satisfied with making NTP status exceptional and pushing it to event queue and not recurring queue (although we are forced to call it recurrently every 10 min), and eventually we can merge this PR, I can cover the code with fat comment explaining this and change the boolean flag.

Reply:
> Yes, if network is not up then we have to wait another 10 minutes. Exactly as all other info messages function (with less interval though). So do not quite understand why NTP status (which has pure informative role on the controller) should be exceptional here.

I see you changed this to pass forcePeriodic=false like for the other into messages, which is what I was asking for.
I see that zedagent calls zedcloudCtx.DeferredEventCtx.KickTimer() when it thinks it makes sense to try to send from the queue. In addition there binary exponential timer (from 1 to 15 minutes) in deferred.go. So this would normally be retried relatively quickly if the network was not connected or there is a short outage.